core/src/main/java/com/linecorp/armeria/client/endpoint/SlowStartAwareEndpointSelectionStrategy.java

```diff
@@ -0,0 +1,135 @@
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.endpoint;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiFunction;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+
+import com.linecorp.armeria.client.ClientFactory;
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.endpoint.WeightedRoundRobinStrategy.EndpointsAndWeights;
+
+import io.netty.util.concurrent.ScheduledFuture;
+
+final class SlowStartAwareEndpointSelectionStrategy implements EndpointSelectionStrategy {
+    private final BiFunction<Endpoint, Duration, Integer> endpointWeighter;
+    private final ClientFactory clientFactory;
+
+    SlowStartAwareEndpointSelectionStrategy(EndpointWeighter endpointWeighter,
+                                                   ClientFactory clientFactory) {
+        this.endpointWeighter = endpointWeighter;
+        this.clientFactory = clientFactory;
+    }
+
+    @Override
+    public SlowStartAwareWeightedRoundRobinSelector newSelector(EndpointGroup endpointGroup) {
+        return new SlowStartAwareWeightedRoundRobinSelector(endpointGroup, endpointWeighter);
+    }
+
+    /**
+     * A weighted round robin select strategy.
+     *
+     * <p>For example, with node a, b and c:
+     * <ul>
+     *   <li>if endpoint weights are 1,1,1 (or 2,2,2), then select result is abc abc ...</li>
+     *   <li>if endpoint weights are 1,2,3 (or 2,4,6), then select result is abcbcc(or abcabcbcbccc) ...</li>
+     *   <li>if endpoint weights are 3,5,7, then select result is abcabcabcbcbcbb abcabcabcbcbcbb ...</li>
+     * </ul>
+     */
+    private final class SlowStartAwareWeightedRoundRobinSelector implements EndpointSelector {
+        private final EndpointGroup endpointGroup;
+        private final AtomicInteger sequence = new AtomicInteger();
+        private volatile EndpointsAndWeights endpointsAndWeights;
+        private final BiFunction<Endpoint, Duration, Integer> endpointWeighter;
+
+        private final Duration slowStartInterval = Duration.ofSeconds(10);
+        private final int maxWeight = 1000;
```

Comment:
I'm not sure if we need to set this because the target weight of each endpoint can be determined by the original weight value of the underlying `EndpointGroup`. What do you think?